implement correct parsing for UpdateItem API

StoneDot opened the pull request:
*Issue #, if available:*

* #92
* #82

*Description of changes:*
This PR implements correct parsing for `UpdateItem` API (`upd` command) using [pest](https://pest.rs/). This also contains the following feature enhancements and bug fixes.

* Support a binary type.
* Support a single quotes for a string type.
* Support list_append and if_not_exists function.
* Support spcifying a list element as a path.
* No longer fail when a expression containing hyphens, commas, indexing, etc.

But, it introduces a breaking change.

* A list type is no longer classified automatically based on its element type. The current behavior is described in [dispatch_jsonvalue_to_attrval function](https://github.com/awslabs/dynein/blob/5eb664559bee22f5e4f7931ed5310d47d9d68f36/src/data.rs#L828-L845)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

ryota-sakamoto commented on src/expression.pest:
```diff
@@ -0,0 +1,119 @@
+
+// You can write attribute name for a`b with a``b like MySQL.
+// https://dev.mysql.com/doc/refman/8.0/en/identifiers.html
+quoted_identifier = @{ ("``" | (!"`" ~ ANY))+ }
```
Would you add README that we can quote attribute name surround by `?

StoneDot replied:
All right!

StoneDot replied:
I added it at https://github.com/awslabs/dynein/pull/132/commits/ad91edcc3807c792f04b54acbb0a5fd264d85cb4.

ryota-sakamoto commented on src/parser.rs:
```diff
@@ -0,0 +1,1466 @@
+}
+
+/// Parse single quoted string as is.
+fn parse_single_quote_literal(str: &str) -> String {
```
Would you add test for parse_single_quote_literal just in case?

StoneDot replied:
I can do it. But, I'm not sure what you want to protect by test. Could you elaborate on this?

ryota-sakamoto replied:
From my opinion, the test is to express specification of the source code. So the purpose of the test for parse_single_quote_literal is that we can understand expected behaviour of parse_single_quote_literal that all of char is raw.

StoneDot replied:
I added the test into the commit https://github.com/awslabs/dynein/pull/132/commits/8c858c7c722e97f4354e933dba51948235768cc5 to manifest the difference in the handling of escapes.

ryota-sakamoto commented on src/parser.rs:
What kind of error do we get when we provide invalid string?
For example, in current version, if we provide incomplete string, panic will occur due to our parser. What will be changed?

```bash
$ dy upd 42 --set 'a = "1'
thread 'main' panicked at 'failed to parse right hand object '"1' into AttributeValue.', src/data.rs:579:29
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

StoneDot replied:
Its panic is expected behavior. Improving error messages needs additional work not related to parsing. Thus, I omit in this PR. After merging this, I will start improvements related to error messages. Actually, providing easy-to-understand errors is a quite difficult problem for pest. Some error messages may become non-sense.  But, at least we will make it a more easy-to-read error message like [pest sample](https://pest.rs/).

ryota-sakamoto reviewed (commented):
Thank you for good enhancement.
I have some comments that we can improve README.md. From my opinion, it is easy for understanding if we write README how to specify --set with some literals and what we supported.

StoneDot commented on README.md:
```diff
@@ -636,23 +640,266 @@ $ dy get 42
+    "ThreeStar": [
+      "Just OK - not that great"
+    ],
+    "metadata": {
```
Thank you! I have fixed it.

StoneDot commented on src/expression.pest:
```diff
@@ -0,0 +1,119 @@
+// String literals
+string_literal = _{ double_quote_literal | single_quote_literal }
+double_quote_literal = @{ "\"" ~ ("\\0" | "\\r" | "\\n" | "\\t" | "\\\\" | "\\\"" | "\\'" | (!"\"" ~ ANY))* ~ "\"" }
+single_quote_literal = @{ "'" ~ (!"\'" ~ ANY)* ~ "'" }
```
All right!

StoneDot replied:
I added a description for string literals at commit [644c86b](https://github.com/awslabs/dynein/pull/132/commits/644c86b5ba2731ed1243dde08fd65ba4134b8f2f). Currently, I refrained from adding a description of binary literals because we have not implemented a display of their value yet.

StoneDot commented:
@ryota-sakamoto I have addressed all the points you mentioned. When you finish the review, please ping me because I would like to rebase and squash the commits into the first one.

ryota-sakamoto commented:
@StoneDot Thank you for addressing some points. It looks good to me.

ryota-sakamoto reviewed (approved):
@StoneDot Thank you for addressing some points. It looks good to me.

StoneDot commented:
Thank you for your review despite the significant changes.
